File: workers/python/src/pilar4/health.py
"""Saúde do canal e kill-switch (Pilar 4 — Python). Fail-closed."""
from __future__ import annotations


def channel_health(
    delivered: int,
    sent: int,
    rejected: int,
    read_rate: float,
    reply_rate: float,
    channel_min: float = 60.0,
    rejection_rate_max: float = 5.0,
) -> dict:
    # Sem dados de envio: nunca autopausa uma campanha nova (fail-safe).
    if sent <= 0:
        return {"score": 0.0, "status": "healthy", "rejection_rate": 0.0, "kill_switch": False, "reasons": ["sem_dados"]}

    delivery = _pct(delivered, sent)
    rejection = _pct(rejected, sent)
    score = (delivery / 100) * 30 + (read_rate / 100) * 25 + (reply_rate / 100) * 25 + max(0.0, 1 - rejection / 100) * 20
    score = round(score, 1)

    reasons: list[str] = []
    if rejection > rejection_rate_max:
        status = "critical"
        reasons.append(f"rejeicao_acima_do_limite:{rejection:.1f}%")
    elif score < channel_min:
        status = "critical"
        reasons.append(f"score_abaixo_do_minimo:{score:.1f}")
    elif score < channel_min + 10:
        status = "warning"
        reasons.append(f"score_proximo_do_minimo:{score:.1f}")
    else:
        status = "healthy"

    return {
        "score": score,
        "status": status,
        "rejection_rate": round(rejection, 1),
        "kill_switch": status == "critical",
        "reasons": reasons,
    }


def _pct(a: int, b: int) -> float:
    return (a / b * 100.0) if b > 0 else 0.0

File: packages/core/src/pilar4/health.ts
/** Saúde do canal e kill-switch (Pilar 4) — fail-closed. */

export type ChannelStatus = "healthy" | "warning" | "critical";

export interface ChannelHealthInput {
  delivered: number;
  sent: number;
  rejected: number;
  readRate: number;
  replyRate: number;
  channelMin: number; // score mínimo (0-100)
  rejectionRateMax: number; // % máximo de rejeição
}

export interface ChannelHealth {
  score: number;
  status: ChannelStatus;
  rejectionRate: number;
  killSwitch: boolean;
  reasons: string[];
}

/** Score 0-100: entrega (30) + leitura (25) + resposta (25) + penalidade de rejeição (20). */
export function channelHealth(input: ChannelHealthInput): ChannelHealth {
  const reasons: string[] = [];
  const delivery = input.sent > 0 ? (input.delivered / input.sent) * 100 : 0;
  const rejection = input.sent > 0 ? (input.rejected / input.sent) * 100 : 0;

  let score = 0;
  score += (delivery / 100) * 30;
  score += (input.readRate / 100) * 25;
  score += (input.replyRate / 100) * 25;
  // rejeição zero → 20 pts; 100% → 0 pts
  score += Math.max(0, 1 - rejection / 100) * 20;

  let status: ChannelStatus = "healthy";
  if (rejection > input.rejectionRateMax) {
    status = "critical";
    reasons.push(`rejeicao_acima_do_limite:${rejection.toFixed(1)}%`);
  } else if (score < input.channelMin) {
    status = "critical";
    reasons.push(`score_abaixo_do_minimo:${score.toFixed(1)}`);
  } else if (score < input.channelMin + 10) {
    // faixa de warning: próximo do mínimo
    status = "warning";
    reasons.push(`score_proximo_do_minimo:${score.toFixed(1)}`);
  }

  const killSwitch = status === "critical";
  return {
    score: Math.round(score * 10) / 10,
    status,
    rejectionRate: rejection,
    killSwitch,
    reasons,
  };
}
